examples/arkode/CXX_serial/ark_kpr_nestedmri.cpp

```diff
@@ -0,0 +1,1979 @@
+  SUNMatrix As        = nullptr; // matrix for slow solver
+  SUNLinearSolver LSs = nullptr; // slow linear solver object
+  if (slowimplicit)
+  {
+    As = SUNDenseMatrix(NEQ, NEQ, sunctx);
+    if (check_ptr((void*)As, "SUNDenseMatrix")) return 1;
+    LSs = SUNLinSol_Dense(y, As, sunctx);
+    if (check_ptr((void*)LSs, "SUNLinSol_Dense")) return 1;
+    retval = ARKodeSetLinearSolver(arkode_mem, LSs, As);
+    if (check_flag(retval, "ARKodeSetLinearSolver")) return 1;
+    retval = ARKodeSetJacFn(arkode_mem, J_s);
+    if (check_flag(retval, "ARKodeSetJacFn")) return 1;
+  }
+  retval = ARKodeSStolerances(arkode_mem, opts.rtol, opts.atol);
+  if (check_flag(retval, "ARKodeSStolerances")) return 1;
+  retval = ARKodeSetMaxNumSteps(arkode_mem, 100000);
+  if (check_flag(retval, "ARKodeSetMaxNumSteps")) return 1;
+  retval = ARKodeSetUserData(arkode_mem, (void*)&opts);
+  if (check_flag(retval, "ARKodeSetUserData")) return 1;
+  if (opts.scontrol != 0)
+  {
+    retval = ARKodeSetAdaptController(arkode_mem, scontrol);
+    if (check_flag(retval, "ARKodeSetAdaptController")) return 1;
+    if (opts.set_h0 != 0)
+    {
+      retval = ARKodeSetInitStep(arkode_mem, opts.hs);
+      if (check_flag(retval, "ARKodeSetInitStep")) return 1;
+    }
+    if (opts.slow_pq == 1)
+    {
+      retval = ARKodeSetAdaptivityAdjustment(arkode_mem, 0);
+      if (check_flag(retval, "ARKodeSetAdaptivityAdjustment")) return 1;
+    }
+    if (!std::isnan(opts.slow_safety))
+    {
+      retval = ARKodeSetSafetyFactor(arkode_mem, opts.slow_safety);
+      if (check_flag(retval, "ARKodeSetSafetyFactor")) return 1;
+    }
+  }
+  else
+  {
+    retval = ARKodeSetFixedStep(arkode_mem, opts.hs);
+    if (check_flag(retval, "ARKodeSetFixedStep")) return 1;
+  }
+
+  //
+  // Integrate ODE
+  //
+
+  // Main time-stepping loop: calls ARKodeEvolve to perform the
+  // integration, then prints results. Stops when the final time
+  // has been reached
+  sunrealtype t     = T0;
+  sunrealtype t2    = T0;
+  sunrealtype dTout = (Tf - T0) / Nt;
+  sunrealtype tout  = T0 + dTout;
+  sunrealtype u, v, w, uerr, verr, werr, uerrtot, verrtot, werrtot, errtot,
+    accuracy;
+  uerr = verr = werr = uerrtot = verrtot = werrtot = errtot = accuracy = ZERO;
+  printf("        t          u          v          w        uerr       verr    "
+         "   werr\n");
+  printf("   "
+         "---------------------------------------------------------------------"
+         "-------\n");
+  printf("  %10.6" FSYM " %10.6" FSYM " %10.6" FSYM " %10.6" FSYM "   %.2" ESYM
+         "   %.2" ESYM "   %.2" ESYM "\n",
+         t, NV_Ith_S(y, 0), NV_Ith_S(y, 1), NV_Ith_S(y, 2), uerr, verr, werr);
+  int Nout = 0;
+  while (Tf - t > 1.0e-8)
+  {
+    // reset reference solver so that it begins with identical state
+    retval = ARKodeReset(arkode_ref, t, y);
+
+    // evolve solution in one-step mode
+    retval = ARKodeSetStopTime(arkode_mem, tout);
```

Comment:
I like most of these suggestions, but I don't want the output file to be inundated with information by plotting every step.  As such, I'll keep `tout` and `dTout` so that we only get a "reasonable" number of outputs, but I'll update the example to follow the other suggestions.

Reply:
This does not need to be addressed for this PR but the stop time is still being set on `arkode_mem`. It's not important for this as an example code but think it could impact the tests you've been running to assess the controller options/settings (given the generality of this example I'm assuming it's being used for that testing as well).

Reply:
This is purposeful.  The goal of the test is to assess how well nested multirate adaptivity performs.  Without a call to set the stop time, the solution returned in `y` could either be the current internal solution (if `t < tout` following the last internal step) or interpolated (if that last step would have overtaken `tout`).  To run a "clean" experiment that assesses **only** the multirate adaptivity error, I must set the stop time in `arkode_mem` to rule out potential interpolation error.

It is a separate question whether we want an example to run such a precise experiment, since our eventual goal with examples is to be more instructional.  To that end, I would be willing to move this file to `test/unit_tests/arkode/CXX_serial` and to retain a dramatically simplified version here.

Reply:
For an example this is fine although we could streamline in the future. My main concern is if you're using this same setup for your tests evaluating the efficiency of different adaptivity methods, then using `SetStopTime` could interfere with the controller by inserting steps limited by the output time not the error control. 

Because `ARKodeEvolve` is being called in `ONE_STEP` mode interpolation won't happen. Removing `SetStopTime` will cause the different setups to evaluate the error at different time points but with the way you've setup the reference solution computation I think that shouldn't be an issue and would still be a fair (maybe better) comparison of the different setups.

Reply:
We **really** need to update our tutorial slides, where we consistently say the opposite about one-step mode: "One-step – take a single internal step and then return control back to the calling program. If this step will overtake `tout` then `y(tout)` is interpolated; otherwise `yn` is returned.​"

I'll revisit this test problem (and example) when working on the MRI adaptivity paper.

Reply:
Agreed, we only fixed this in the ARKODE docs about a year ago (#362).